.chloggen/elasticsearchexporter_sync-bulk-indexer-flush-bytes.yaml

```diff
@@ -0,0 +1,31 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: elasticsearchexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Respect `flush::bytes` in sync bulk indexer, `flush::bytes` measures uncompressed size, change default `batcher::max_size_items` to `0`
```

Comment:
Do we really need to remove the `batcher::max_size_items`? Can't they work together? Something like batch should not have more than `batcher::max_size_items` AND size greater than `flush::bytes`?

Reply:
> They can work together, but I don't think max_size_items provide any value at all, after enforcing

Might be missing something here, don't they still provide an upper bound on the max items size? 5MB max bytes and 10,000 max size items default would give us 0.5KB per item a bit too low but not impossible I think. IMO, both size and number would be factor worth considering for designing the payload since the processing of this payload will also take these 2 factors in account.

Reply:
I don't mind reverting the max_size_items change, but I just don't see the value of a default. There may be minor implications on perf, and imo having max_size_items may just lead to less effective batching, if the items are all very small. e.g. If a 5MB bulk request can accommodate 15k items, I don't see why we should artificially limit max_size_items to a default of 10k, which results in a 10k items request and a 5k items request.

Reply:
> I don't see why we should artificially limit max_size_items to a default of 10k, which results in a 10k items request and a 5k items request

IIRC, we introduced the max size items as a way to implicitly limit size since batcher didn't provide us an option so I don't have a strong opinion on keeping/removing it but, IMO, the number of actions that an endpoint performs for a request is an important optimization factor and I think having an upper limit here is as important as the byte size limit. Different workloads can perform differently for different payloads (w.r.t. both size and item count).

Reply:
Actually, this reminds me that setting max_size_items may actually split metrics that should be grouped in the same document, which will lead to a duplication for metrics in TSDB. I haven't seen that yet just because I haven't tested 8.16 otel mapping mode + batcher, but it sounds very likely that it will cause metrics to be rejected.

On the other hand, if we use `flush::bytes` after batching, the documents are already grouped correctly, and chunking the bulk request at this level will not mess with metrics grouping. Is this argument good enough to remove max_size_items?

Reply:
> this reminds me that setting max_size_items may actually split metrics that should be grouped in the same document, which will lead to a duplication for metrics in TSDB

Ah, good point. Will this also have an impact in the future if we were to move to the batcher when the batcher has feature for flush bytes configuration?

> Is this argument good enough to remove max_size_items?

Yup. Should we remove this feature altogether then (in a different PR)? Or at least put up a warning?

Reply:
> Will this also have an impact in the future if we were to move to the batcher when the batcher has feature for flush bytes configuration?

Yes. I assume in the future when byte-based batching is supported by batcher, we won't be able to use `max_size_bytes`, but only `min_size_bytes`, for the same reason.

> Yup. Should we remove this feature altogether then (in a different PR)? Or at least put up a warning?

Good call. I'll add a warning in the README.

Reply:
> Yes. I assume in the future when byte-based batching is supported by batcher, we won't be able to use max_size_bytes, but only min_size_bytes, for the same reason.

Not sure I see how min size bytes will be okay but without the implementation we can probably not be sure either way. IIUC, if we were to implement the [Request](https://github.com/open-telemetry/opentelemetry-collector/blob/c6799fb4ddf41783693886f5238787c60221b079/exporter/internal/request.go#L15) and the [RequestErrorHandler](https://github.com/open-telemetry/opentelemetry-collector/blob/c6799fb4ddf41783693886f5238787c60221b079/exporter/internal/request.go#L45) interface then we would have full control over how we design the export payload. It would probably require us to do a much better accounting of the size when converting OTel data model to ES data model.

All that being said, I think this is a problem for future and let's have this discussion at a later point.

Reply:
> Not sure I see how min size bytes will be okay

The problem with max size (items or bytes) is that it will force split something that being to the same Request. Min size doesn't split things, which means everything in the same Request will be considered during grouping.

Reply:
I get that but we will be implementing the logic for splitting. So, theoretically at least, we will have the capability to split it in a way that avoids splitting the metrics across documents. This will require better size estimation of items before they are added to the batch but it should be possible to do I think.